TAO/TAO_IDL/fe/idl.ypp

```diff
@@ -3889,6 +3892,152 @@ enumerator :
         }
         ;
 
+map_type_spec :
+        IDL_MAP
+        {
+          idl_global->scopes ().push (0);
+        }
+        '<'
+        map_type
+        ','
+        map_type
+        '>'
+        {
+
+          AST_Map *map = 0;
+          Decl_Annotations_Pair *key_type = $4;
+          Decl_Annotations_Pair *val_type = $6;
+
+          /*
+           * Remove sequence marker from scopes stack.
+           */
+          if (idl_global->scopes ().top () == 0)
+            {
+             idl_global->scopes ().pop ();
+            }
+
+          UTL_Scope *s = idl_global->scopes ().top_non_null ();
+
+          /*
+           * Create a node representing a sequence.
+           */
+          if (key_type && val_type)
+            {
+              AST_Type *ktp = dynamic_cast<AST_Type*> (key_type->decl);
+              AST_Type *vtp = dynamic_cast<AST_Type*> (val_type->decl);
+
+              if (ktp == 0 || vtp == 0)
+                {
+                  ; // Error will be caught in FE_Declarator.
+                }
+              else
+                {
+                  Identifier id ("map");
+                  UTL_ScopedName sn (&id, 0);
+                  ACE_CDR::ULong bound = 0UL;
+
+                  map =
+                    idl_global->gen ()->create_map (
+                        idl_global->gen ()->create_expr (
+                                                bound,
+                                                AST_Expression::EV_ulong
+                                              ),
+                        ktp,
+                        vtp,
+                        &sn,
+                        s->is_local (),
+                        s->is_abstract ()
+                      );
+                  map->key_type_annotations (*key_type->annotations);
+                  map->value_type_annotations (*val_type->annotations);
+
+                  idl_global->err ()->anonymous_type_diagnostic ();
+                }
+            }
+
+          delete key_type->annotations;
+          delete val_type->annotations;
+          $$ = map;
+        }
+        | IDL_MAP
+        /* {
+          Adding this causes error: 1 reduce/reduce conflict [-Werror=conflicts-rr]
+          idl_global->scopes ().push (0);
+        } */
```

Comment:
Yea, that protects the compiler against using ambiguous grammar that would otherwise wouldn't show up as an issue unless you try to use the particular grammar. I think in this case when it's encountering `IDL_MAP` it can't decide between going with the unbounded and bounded variants of map. Before it was okay, but adding the code to run with `IDL_MAP` probably makes it impossible for the parser to undo this if the whole rule doesn't work out and it wanted to check for other rules.

I think that's why in sequence this is a separate rule as `seq_head`. I suggest replicating that.

Reply:
This is what I had initially started doing, however I ran into issues seq_head returns(?) the annotation+decl type for the base type of the sequence however I wasn't sure how to do this with map since it would need to return two types back into the main map_type_spec parsing. Would I create a struct that groups two annotation+decl types together or is there a better way?

Reply:
That's a good question. Making a new struct with 2 `Decl_Annotations_Pair` is probably good enough. An alternative might be to use something more generic like `std::vector<Decl_Annotations_Pair>`, but that might be overkill.

Reply:
Wait actually if I can add C++ std things, would a std::pair be fine then

Reply:
> Wait actually if I can add C++ std things, would a std::pair be fine then

Yes, that would work too. The reason there's not a lot of `std` stuff is because the code predates it, but it's fine to use `std` stuff now.

Reply:
I ended up going with a new struct Decl_Annotations_Pair_Pair (I can change the name if it's a bit to confusing), as both std::pair and std::vector was complaining about either non trivial default constructors or copy constructors, I ended up giving up on that path